### PR TITLE
Option to delete file on record deletion

### PR DESCRIPTION
**Pull request by jorisvaesen**

### CODE REVIEW

**Comment by josegonzalez** [comment]
Just fyi, I only merge changes that do not decrease test coverage (including PHPCS!) and have the requisite docs. Happy to work towards that though :)

File: src/File/Writer/DefaultWriter.php [hunk 1]
```diff
@@ -85,6 +85,23 @@ public function write(array $files)
     }
 
     /**
+     * Deletes a set of files to an output
+     *
+     * @param array $files the files being written out
+     * @return array array of results
+     */
+    public function delete(array $files)
```
Comment (josegonzalez) [inline src/File/Writer/DefaultWriter.php]: You need a test for this method as well.

File: tests/TestCase/Model/Behavior/UploadBehaviorTest.php [hunk 2]
```diff
@@ -201,6 +201,15 @@ public function testBeforeSaveOk()
         $this->assertNull($behavior->beforeSave(new Event('fake.event'), $this->entity, new ArrayObject));
     }
 
+    public function testAfterDelete()
```
Comment (josegonzalez) [inline tests/TestCase/Model/Behavior/UploadBehaviorTest.php]: You need both a successful and an unsuccessful test run.

**Comment by josegonzalez** [comment]
This works now?

**Comment by josegonzalez** [comment]
@jorisvaesen does this work according to how you'd like it now?

**Comment by jorisvaesen** [comment]
I did some testing and it seems to work as expected.

**Comment by josegonzalez** [comment]
Thanks for being patient!

**Comment by josegonzalez** [comment]
@jorisvaesen added you as a maintainer. Don't abuse your new role of course! I'm looking forward to your help in maintaining this plugin, especially considering the nice features you've worked on :)

**Comment by KingLoui** [comment]
maybe it is a silly question but how can I delete additionally generated thumbnails with this feature? I thought I would just overwrite the delete method in the writerinterface but the passed array $files only has the path and not the filename, so I cant reference my thumbnail from there.

**Comment by KingLoui** [comment]
I just did a quick and dirty solution to show the issue.

MyOwnFilterWriter.php
```
public function delete(array $files)
{
        
    $filesystem = $this->getFilesystem($this->field, $this->settings);
    $results = [];
    $results[] = $this->deletePath($filesystem, $this->entity['file_dir'] . $this->entity['file']);
    $results[] = $this->deletePath($filesystem, $this->entity['file_dir'] . 'thumbnail-' . $this->entity['file']);

    return $results;
}
```
my solution obviously doesn't work when you have multiple files per entity but my MyOwnWriter is specific for my database so in my case there wont be an issue.

**Comment by josegonzalez** [comment]
@KingLoui since we cannot tell what your original urls will be - that isn't stored - your method is more or less what I would do.

```
public function delete(array $files)
{
    $files[] = $this->entity['file_dir'] . 'thumbnail-' . $this->entity['file']
    return parent::delete($files);
}
```

The above might be better.

**Comment by jorisvaesen** [comment]
For automating this, I was thinking to use the transformer.

Create an extra function, let's say `beforeTransform()` where thumbnails can be created.
And let `transform()` only return the array of files which can later be used by the writer to delete all files (including thumbnails).

**Comment by josegonzalez** [comment]
Maybe, though I'd have to see implementation/docs first. Also, requiring this in the interface would be a BC break, just fyi.

**Comment by jorisvaesen** [comment]
Indeed.
For 3.x the writers `delete` function should be overridden (like you said before) but maybe it could be an option for 4.x ?

**Comment by josegonzalez** [comment]
Sure.

**Pull request merged**